examples/src/bin/gif_paintable.rs

```diff
@@ -0,0 +1,305 @@
+// crate to read a gif file and transform it to a Vec<Frame>
+// Which are then rendered by the paintable at snapshot
+// A tick callback that calls invalidate_contents on the paintable
+// to render a new frame
+use glib::clone;
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use gtk::{gdk, gio, glib, graphene};
+use image::{gif::GifDecoder, AnimationDecoder};
+use std::env::args;
+use std::io::Cursor;
+
+pub struct Frame {
+    pub width: i32,
+    pub height: i32,
+    pub texture: gdk::Paintable,
+}
+
+impl Frame {
+    pub fn new(f: image::Frame) -> Self {
+        let image = f.into_buffer();
+
+        let samples = image.into_flat_samples();
+        let (stride, width, height) = samples.extents();
+        let width = width as i32;
+        let height = height as i32;
+
+        let bytes = glib::Bytes::from(samples.as_slice());
+
+        let texture =
+            gdk::MemoryTexture::new(width, height, gdk::MemoryFormat::R8g8b8a8, &bytes, stride);
+        Frame {
+            width,
+            height,
+            texture: texture.upcast(),
+        }
+    }
+}
+
+mod imp {
+    use super::*;
+    use glib::subclass;
+    use std::cell::{Cell, RefCell};
+
+    pub struct GifPaintable {
+        pub frames: RefCell<Vec<Frame>>,
+        pub next_frame: RefCell<Option<gdk::Paintable>>,
+        pub width: Cell<i32>,
+        pub height: Cell<i32>,
+        pub current_idx: Cell<usize>,
+    }
+
+    impl ObjectSubclass for GifPaintable {
+        const NAME: &'static str = "GifPaintable";
+        type Type = super::GifPaintable;
+        type ParentType = glib::Object;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+
+        glib::object_subclass!();
+
+        fn type_init(type_: &mut subclass::InitializingType<Self>) {
+            type_.add_interface::<gdk::Paintable>();
+        }
+
+        fn new() -> Self {
+            Self {
+                next_frame: RefCell::new(None),
+                frames: RefCell::new(vec![]),
+                width: Cell::new(0),
+                height: Cell::new(0),
+                current_idx: Cell::new(0),
+            }
+        }
+    }
+
+    impl ObjectImpl for GifPaintable {
+        fn constructed(&self, obj: &Self::Type) {
+            self.parent_constructed(obj);
+        }
+
+        fn dispose(&self, _obj: &Self::Type) {}
+    }
+
+    impl PaintableImpl for GifPaintable {
+        fn get_intrinsic_height(&self, _paintable: &Self::Type) -> i32 {
+            self.height.get()
+        }
+
+        fn get_intrinsic_width(&self, _paintable: &Self::Type) -> i32 {
+            self.width.get()
+        }
+
+        fn snapshot(
+            &self,
+            _paintable: &Self::Type,
+            snapshot: &gdk::Snapshot,
+            width: f64,
+            height: f64,
+        ) {
+            if let Some(texture) = self.next_frame.borrow_mut().take() {
```

Comment:
Should it not continue returning the same frame until it's time for the following one?

Reply:
Indeed,